helix-lock/src/main/java/org/apache/helix/lock/helix/ZKDistributedNonblockingLock.java

```diff
@@ -122,9 +210,91 @@ public void close() {
+          .getRequestorPriority()) {
+        LOG.debug("We do not need to handle this data change");
+      } else {
+        _lockListener.onCleanupNotification();
+        // read the lock information again to avoid stale data
+        readData = _baseDataAccessor.get(dataPath, stat, AccessOption.THROW_EXCEPTION_IFNOTEXIST);
+        // If we are still the lock owner, clean the lock owner field.
+        if (lockInfo.getOwner().equals(_userId)) {
+          ZNRecord znRecord = new ZNRecord(readData);
+          znRecord.setSimpleField(LockInfo.LockInfoAttribute.OWNER.name(),
+              LockConstants.DEFAULT_OWNER_TEXT);
+          znRecord.setSimpleField(LockInfo.LockInfoAttribute.MESSAGE.name(),
+              LockConstants.DEFAULT_MESSAGE_TEXT);
+          znRecord.setLongField(LockInfo.LockInfoAttribute.TIMEOUT.name(),
+              LockConstants.DEFAULT_TIMEOUT_LONG);
+          znRecord.setIntField(LockInfo.LockInfoAttribute.PRIORITY.name(),
+              LockConstants.DEFAULT_PRIORITY_INT);
+          znRecord.setLongField(LockInfo.LockInfoAttribute.WAITING_TIMEOUT.name(),
+              LockConstants.DEFAULT_WAITING_TIMEOUT_LONG);
+          znRecord.setLongField(LockInfo.LockInfoAttribute.CLEANUP_TIMEOUT.name(),
+              LockConstants.DEFAULT_CLEANUP_TIMEOUT_LONG);
+          _baseDataAccessor
+              .update(_lockPath, new ZNRecordUpdater(znRecord), AccessOption.PERSISTENT);
+        } else {
+          LOG.info("We are not current lock owner");
+        }
+      }
+    } // We are the current requestor
+    else if (lockInfo.getRequestorId().equals(_userId)) {
+      // In case the owner field is empty, it means previous owner has finished cleanup work.
+      if (lockInfo.getOwner().equals("NONE")) {
+        ZNRecord znRecord = composeNewOwnerRecord();
+        LockUpdater updater = new LockUpdater(new LockInfo(znRecord));
+        _baseDataAccessor.update(_lockPath, updater, AccessOption.PERSISTENT);
+        onAcquiredLockNotification();
+      } else {
+        LOG.info("We do not need to handle this data change");
+      }
+    } // If we are waiting for the lock, but find we are not the requestor any more, meaning we
+    // are preempted by an even higher priority request
+    else if (!lockInfo.getRequestorId().equals("NONE") && !lockInfo.getRequestorId().equals(_userId)
+        && _isPending) {
+      onDeniedPendingLockNotification();
+    }
+  }
+
+  /**
+   * call back called when the lock is acquired
+   */
+  public void onAcquiredLockNotification() {
+    synchronized (ZKDistributedNonblockingLock.this) {
+      _isLocked = true;
+      ZKDistributedNonblockingLock.this.notify();
+    }
+  }
+
+  /**
+   * call back called when the pending request is denied due to another higher priority request
+   */
+  public void onDeniedPendingLockNotification() {
+    synchronized (ZKDistributedNonblockingLock.this) {
+      _isLocked = false;
```

Comment:
Looks like a lock can only be in one state at a time. I suggest merging these 2 booleans into an enum for the cleanness of the code and the logic.